File: runme/mcgen/local/datan/exasrc/e6mtx.c
#include <stdio.h>
#include "datsrc.h"

/* Table of constant values */

static integer c__3 = 3, c__4 = 4;

main(void)
{
    /* Initialized data */

    static integer m, n;
    static double b[9] /* was [3][3] */ = { 1.,3.,5.,7.,11.,13.,17.,19.,23. },
                  c[6] /* was [3][2] */ = { 1.,1.,1.,1.,1.,1. },
                  a[9] /* was [3][3] */, r[9] /* was [3][3] */,
                  s[9] /* was [3][3] */, u[9] /* was [3][3] */;

/* demonstrates use of Cholesky routines */
/* identify program to user */
    printf("%s\n", "Program E6MTX demonstrates use of");
    printf("%s\n", "MTXCHL,MTXCHI and MTXCHM");
    printf("%s\n\n", "for Cholesky decomposition and inversion.");
/* construct symmetric, positive definite matrix X */
    n = 3;
    m = 2;
    mtxmat_(b, b, a, &n, &n, &n);
    printf("%s\n", "Output produced by program E6MTX.");
    printf("%s\n", "------------------------------------------");
/* write input data */
    printf("%s\n", "A =");
    mtxwrt_(a, &n, &n);
    printf("\n%s\n", "C =");
    mtxwrt_(c, &n, &m);
    printf("\n%s\n\n", "------------------------------------------");
/* demonstrate MTXCHL */
    mtxchl_(a, u, &n);
    printf("%s\n", "CALL MTXCHL(A,U,N) yields U = ");
    mtxwrt_(u, &n, &n);
    printf("\n");
/* perform check by multiplying U with its transposed */
    mtxmat_(u, u, r, &n, &n, &n);
    printf("%s\n", "check of decomposition: R = U^T U =");
    mtxwrt_(r, &n, &n);
    printf("\n");
/* demonstrate MTXCHM */
    mtxchm_(u, c, r, &n, &m);
    printf("%s\n", "CALL MTXCHM(U,C,R,N,M)  yields  R =");
    mtxwrt_(r, &n, &m);
    printf("\n");
/* demonstrate MTXCHI */
/* (first save original matrix A) */
    mtxtra_(a, s, &n, &n);
    mtxchi_(a, u, &n);
    printf("%s\n", "CALL MTXCHI(A,U,N) yields  A =");
    mtxwrt_(a, &n, &n);
    printf("\n");
/* perform check by multiplying original matrix A with inverse */
    mtxmlt_(s, a, r, &n, &n, &n);
    printf("%s\n", "check of inversion: R = A**(-1) A =");
    mtxwrt_(r, &n, &n);
    printf("\n");
    return 0;
} /* main */

File: runme/mcgen/local/datan/exasrc/e6mtx.f
      PROGRAM E6MTX
C demonstrates use of Cholesky routines
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      DIMENSION B(3,3),C(3,2),A(3,3),U(3,3),R(3,3),S(3,3)
      DATA B/1.D0,3.D0,5.D0,7.D0,11.D0,13.D0,17.D0,19.D0,23.D0/
      DATA C /1.D0,1.D0,1.D0,1.D0,1.D0,1.D0/
C identify program to user
      WRITE(*,*)' Program E6MTX demonstrates use of'
      WRITE(*,*)' MTXCHL,MTXCHI and MTXCHM'
      WRITE(*,*)' for Cholesky decomposition and inversion'
      WRITE(*,*)' '
C construct symmetric, positive definite matrix X
      N=3
      M=2
      CALL MTXMAT(B,B,A,N,N,N)
      WRITE(*,*)' Output produced by program E6MTX'
      WRITE(*,'(A)')' ------------------------------------------'
C write input data
      WRITE(*,*)' A ='
      CALL MTXWRT(A,N,N)
      WRITE(*,*)' '
      WRITE(*,*)' C ='
      CALL MTXWRT(C,N,M)
      WRITE(*,*)' '
      WRITE(*,'(A)')' ------------------------------------------'
      WRITE(*,*)' '
C demonstrate MTXCHL
      CALL MTXCHL(A,U,N)
      WRITE(*,'(A)')' CALL MTXCHL(A,U,N) yields U = '
      CALL MTXWRT(U,N,N)
      WRITE(*,*)' '
C perform check by multiplying U with its transposed
      CALL MTXMAT(U,U,R,N,N,N)
      WRITE(*,*)' check of decomposition: R = U^T U ='
      CALL MTXWRT(R,N,N)
      WRITE(*,*)' '
C demonstrate MTXCHM
      CALL MTXCHM(U,C,R,N,M)
      WRITE(*,'(A)')' CALL MTXCHM(U,C,R,N,M)  yields  R ='
      CALL MTXWRT(R,N,M)
      WRITE(*,*)' '
C demonstrate MTXCHI
C (first save original matrix A)
      CALL MTXTRA(A,S,N,N)
      CALL MTXCHI(A,U,N)
      WRITE(*,'(A)')' CALL MTXCHI(A,U,N) yields  A ='
      CALL MTXWRT(A,N,N)
      WRITE(*,*)' '
C perform check by multiplying original matrix A with inverse
      CALL MTXMLT(S,A,R,N,N,N)
      WRITE(*,*)' check of inversion: R = A**(-1) A ='
      CALL MTXWRT(R,N,N)
      WRITE(*,*)' '
      END
